Enable re-encoding of segments

Bouncner opened the pull request:
This PR introduces re-encoding to chunks/segments. Instead of passing a `ValueSegment` to the encoders, an `AnySegmentIterable` is passed.
Statistics are only built once.

__What has been enabled by this PR:__
* Re-encoding from an encoded segment to an `Unencoded` one is now possible.
* Loading binary files (e.g., from dictionary segment exports) and re-encoding them  (e.g., `-e LZ4`) is now possible.

__Performance:__
Performance got slightly faster (except for std::strings), but that's barely noticeable for most users. Fixed string encoding is much faster as one loop less over the data is performed.
```
----------------------------------------------------------------------
Benchmark                               Time           CPU Iterations
----------------------------------------------------------------------
BM_EncodingPerf_Integer_old      15627849 ns   15627568 ns        896
BM_EncodingPerf_Integer_new      14100860 ns   14100774 ns        992
BM_EncodingPerf_String_old      227411133 ns  227389640 ns         47
BM_EncodingPerf_String_new      254255025 ns  254225617 ns         60
BM_EncodingPerf_FSString_old    725246251 ns  725228876 ns         20
BM_EncodingPerf_FSString_new    247519962 ns  247511163 ns         58
BM_EncodingPerf_LZ4Integer_old   15705026 ns   15704350 ns        888
BM_EncodingPerf_LZ4Integer_new   14179989 ns   14179962 ns        988
BM_EncodingPerf_FoRInteger_old   15624515 ns   15624161 ns        897
BM_EncodingPerf_FoRInteger_new   14124000 ns   14123530 ns        994
```

Fixes #1368.

mrks commented on src/lib/storage/lz4_segment/lz4_encoder.hpp:
```diff
@@ -168,35 +165,38 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
+
+      null_values.resize(segment_size);
+      offsets.resize(segment_size);
+      sample_sizes.resize(segment_size);
```
Legacy, but maybe `string_length_samples` is better?

Bouncner replied:
Changed to `string_samples_lengths`.

mrks commented on src/lib/storage/mvcc_data.cpp:
```diff
@@ -26,8 +26,12 @@ void MvccData::shrink() {
+  _is_shrunk = true;
 }
 
+bool MvccData::is_shrunk() { return _is_shrunk; }
```
Did you check if shrinking has any cost on a vector that is already shrunk?

Bouncner replied:
On a 100'000 size_t vector, shrink() takes ~2000ns and the following shrink()'s take <200ns.
So there is an overhead, but probably not enough. Will remove it.

Code:
```
BENCHMARK_F(MicroBenchmarkBasicFixture, TBBShrink)(benchmark::State& state) {
  tbb::concurrent_vector<size_t> values;

  for (auto i = size_t{0}; i < 100'000; ++i) {
    values.push_back(i);
  }

  auto start_time = std::chrono::high_resolution_clock::now();
  values.shrink_to_fit();
  auto end_time = std::chrono::high_resolution_clock::now();
  std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start_time).count() << std::endl;;

  for (auto i = size_t{0}; i < 5; ++i) {
    start_time = std::chrono::high_resolution_clock::now();
    values.shrink_to_fit();
    end_time = std::chrono::high_resolution_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start_time).count() << std::endl;;
  }
}
```

mrks commented on src/lib/storage/segment_encoding_utils.cpp:
```diff
@@ -52,4 +51,22 @@ std::shared_ptr<BaseEncodedSegment> encode_segment(EncodingType encoding_type, D
+    case CompressedVectorType::FixedSize4ByteAligned:
+    case CompressedVectorType::FixedSize2ByteAligned:
+    case CompressedVectorType::FixedSize1ByteAligned:
+      return VectorCompressionType::FixedSizeByteAligned;
```
Uargh. Introducing "parent types" makes #1286 even more necessary. `VectorCompressionAlgorithm` and `VectorCompressionType`?

Bouncner replied:
Copied your recommendation to the issue.

mrks commented on src/test/storage/chunk_encoder_test.cpp:
```diff
@@ -146,4 +151,26 @@ TEST_F(ChunkEncoderTest, EncodeMultipleChunksUsingSameEncoding) {
 }
 
+TEST_F(ChunkEncoderTest, ReencodingTable) {
+  const auto chunk_encoding_specs =
```
Comment on what this holds?

Bouncner replied:
Added `// Encoding specifications which will be applied one after another to the chunk.`

mrks commented on src/test/storage/encoded_segment_test.cpp:
```diff
@@ -171,7 +185,7 @@ TEST_P(EncodedSegmentTest, SequentiallyReadNotNullableIntSegment) {
 TEST_P(EncodedSegmentTest, SequentiallyReadNullableIntSegment) {
   auto value_segment = create_int_with_null_value_segment();
-  auto base_encoded_segment = encode_value_segment(DataType::Int, value_segment);
+  auto base_encoded_segment = encode_base_segment(DataType::Int, value_segment);
```
Just "encode_segment"?

Bouncner replied:
Did that, but for some reason clang does no longer find the eventual `encode_segment` method in the utils, even though the signature differ. Might be related to the other problem where I use `::opossum::`?

Any ideas what I am doing wrong?

mrks replied:
Hard to check without access to a compiler. Not a.big thing anyway.

mrks commented on src/test/storage/encoded_segment_test.cpp:
```diff
@@ -271,4 +285,55 @@ TEST_P(EncodedSegmentTest, SequentiallyReadEmptyIntSegment) {
+            EXPECT_EQ(value_segment_it->is_null(), encoded_segment_it->is_null());
+
+            if (!value_segment_it->is_null()) {
+              EXPECT_EQ(value_segment_it->value(), encoded_segment_it->value());
```
Simply add the segments to tables and check for their equality?

Bouncner replied:
If I remember correctly, table equality checks are rather expensive.

mrzzzrm replied:
just use check_table_equals, it produces much more readable test failures.

Bouncner replied:
Haven't found a check for segment equality. So I need to construct to tables, append a chunk with that segment and then check for equality. Right?

mrks replied:
Yes, but a function `check_segments_equal` would be the same effort while being more reusable.

Bouncner replied:
What do you think about #e71342f? What bothers me a little bit is that the result prints "Tables not equal" in case of a failure. But I'd like to avoid bloating up the .cpp file just to have a special version of the comparison method for segments.

mrks replied:
No # in front of commits. e71342f (so that I have something to click on)

mrks replied:
That’s the general idea, yes. Just two things: First, you are ignore TypeCmpMode by requiring the types to be the same in any case (have a look at how that parameter is used later). Second, you would store ReferenceSegments in a data table.

Bouncner replied:
Changed both issues (there is a "full version" now and two specializations, similar to `EXPECT_TABLE_EQ`). Still the "issue" with "Tables not equal".

Bouncner replied:
@mrzzzrm : I'd like to close most issues here. Apart from the compiler issue (`::opossum::`) the issue with the `check_segments_equal` remains. When it fails, it prints something like "Tables are not equal". Any ideas or is it not worth changing it?

mrks commented on src/test/storage/encoded_segment_test.cpp:
```diff
@@ -271,4 +285,55 @@ TEST_P(EncodedSegmentTest, SequentiallyReadEmptyIntSegment) {
+  };
+
+  // Use the row_count used for frame of reference segments.
+  auto value_segment = create_int_with_null_value_segment(row_count(EncodingType::FrameOfReference));
```
Please use a paremeterized test instead.

Bouncner replied:
Not sure what you mean. Do you mean using `GetParam()` and `TEST_P`?

mrks replied:
yes

Bouncner replied:
The tests are parameterized to take a encoding configuration and test this configuration. But the purpose of the reencoding test is to reencode between multiple configurations and check that actual data remains the same.

mrks replied:
Ok

mrks commented on src/test/storage/encoded_string_segment_test.cpp:
```diff
@@ -294,4 +302,54 @@ TEST_P(EncodedStringSegmentTest, SequentiallyReadNullableStringSegmentWithShuffl
   });
 }
 
+TEST_F(EncodedStringSegmentTest, SegmentReencoding) {
```
Same as the non-strings

Bouncner commented on src/lib/storage/base_segment_encoder.hpp:
```diff
@@ -131,14 +131,17 @@ class SegmentEncoder : public BaseSegmentEncoder {
 
-    const auto value_segment = std::dynamic_pointer_cast<const ValueSegment<ColumnDataType>>(base_value_segment);
-    Assert(value_segment, "Value segment must have passed data type.");
+    // Obtain allocator in case of value segment.
```
Changed to
```
    // Obtain allocator of segment to encode. Currently, only ValueSegment store the used allocator.
    // TODO(anyone): obtain allocator for all types of segment, when available.
```

Ok?

mrks replied:
yepp

Bouncner commented on src/lib/storage/chunk_encoder.cpp:
```diff
@@ -12,41 +12,104 @@
+
+    const auto unencoded_segment = std::dynamic_pointer_cast<const ValueSegment<ColumnDataType>>(segment);
+    if (unencoded_segment && encoding_spec.encoding_type == EncodingType::Unencoded) {
+      return_segment = segment;
```
Renamed to `result`.

Bouncner commented on src/lib/storage/chunk_encoder.cpp:
```diff
@@ -12,41 +12,104 @@
+      }
+    }
+
+    const auto unencoded_segment = std::dynamic_pointer_cast<const ValueSegment<ColumnDataType>>(segment);
```
I am checking whether the given BaseSegment is of type ValueSegment. Is that wrong? Using `resolve_*` would unnecessarily increase compile times, wouldn't it?

mrks replied:
Not this line in particular, but the block started here.

Bouncner replied:
Will add a comment. But this cast to check for the type shouldn't create/copy any data, right?

mrks replied:
No.

Bouncner commented on src/lib/storage/dictionary_segment/dictionary_encoder.hpp:
```diff
@@ -28,20 +29,46 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
+        const auto segment_item = *segment_it;
+        if (!segment_item.is_null()) {
+          const auto segment_value = segment_item.value();
+          values.push_back(segment_value);
```
Added comments.

Bouncner commented on src/lib/storage/dictionary_segment/dictionary_encoder.hpp:
```diff
@@ -28,20 +29,46 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
+    std::vector<bool> null_values;
+
+    auto max_string_length = size_t{0};
+    std::set<T> unique_values;
```
I am currently looking why clang fails (it does not on my Mac or nemea). After that, I will reevaluate with GCC.

But my current assumption is that obtaining values via an erased iterator hides the overhead of adding the values to the map. After that, obtaining unique elements in a sorted order is fast using the map.
That's very different from previous measurements were the input was a std::vector directly holding the values in a consecutive chunk of memory.

mrks replied:
Did you test inserting into a vector, sorting it, and making it unique? If it doesn‘t make a difference, it doesn‘t matter.

Bouncner replied:
Yes, did test that. Was slower in most cases (for strings clearly slower).

Bouncner replied:
Thanks for bringing that up! The initially better performance was using std::unsorted_set, which does not provide a sorted order. With sorting it is slower than using std::set. And std::set is slower than the initial approach. I also tested using `std::unique_copy` but it did not show any performance improvements.

Bouncner replied:
I updated the PR information with the current numbers.

Bouncner commented on src/lib/storage/lz4_segment/lz4_encoder.hpp:
```diff
@@ -168,35 +165,38 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
      */
-    auto offsets = pmr_vector<uint32_t>{alloc};
-    offsets.resize(num_elements);
+    auto offsets = pmr_vector<uint32_t>{allocator};
```
Shouldn't change anything. Binary caching still only works for values and dictionary segments, since the current benchmark do only support an initial encoding configuration, re-encoding doesn't happen anyways.

mrks replied:
What if the stored binary table is different from the requested one?

Bouncner replied:
Ahh, I see your point. That should work just fine now. I will check if there are some quirks.

Bouncner replied:
Works like a charm: https://gist.github.com/Bouncner/2d134821662626047cb8d629660e504c

Bouncner commented on src/lib/storage/dictionary_segment/dictionary_encoder.hpp:
```diff
@@ -53,87 +80,41 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
+    auto encoded_attribute_vector =
+        compress_vector(attribute_vector, SegmentEncoder<DictionaryEncoder<Encoding>>::vector_compression_type(),
+                        allocator, {max_value});
+    auto dictionary_sptr = std::allocate_shared<U>(allocator, std::move(dictionary));
```
Any recommendation.

mrks replied:
Why can’t we allocate with the correct allocator in the first place?

Bouncner replied:
I never really tried to understand the the meaning of this allocate_shared and thought is allocation magic. I will take a look at it.

Bouncner replied:
Apparently the reason is that shared_pointers are passed to the dictionary segment's constructor. I do not see why shared_pointers are passed to be honest. Do you know?

mrks replied:
`allocate_shared` is like `make_shared`, but with an additional allocator passed in. Here, we are creating the dictionary with the default allocator and then copy (not move) it to memory allocated with the correct allocator. Does that copy make sense?

mrks replied:
Also, in case there was any confusion: No need to get rid of the `shared_ptr`, the original comment was about the „sptr“ prefix

Bouncner replied:
Not a big fan of all these shared_pointers. But I changed the code accordingly with d3ef430.

I allocate directly a shared object, to be able to use the allocator. From my understanding, first creating the object unshared and then making it shared cannot be done with a custom allocator, right?

Bouncner commented on src/lib/storage/chunk_encoder.cpp:
```diff
@@ -12,41 +12,104 @@
 
 namespace opossum {
 
+std::shared_ptr<BaseSegment> ChunkEncoder::encode_segment(const std::shared_ptr<BaseSegment>& segment,
```
Added a check and a fail.

Bouncner commented on src/lib/storage/dictionary_segment/dictionary_encoder.hpp:
```diff
@@ -28,20 +29,46 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
+      }
+    });
+
+    pmr_vector<T> dictionary(std::make_move_iterator(unique_values.cbegin()),
```
I don't get your comment. We do use the correct allocator and we don't have the unique (sorted) values before this point.

mrks replied:
Ah, the comment was under the assumption that `unique_values` is still a vector. Nevermind.

mrks commented on src/lib/storage/chunk_encoder.cpp:
```diff
@@ -12,41 +12,114 @@
+        for (; it != end; ++it) {
+          const auto segment_item = *it;
+          const auto is_null = segment_item.is_null();
+          null_values.push_back(is_null);
```
Why do you use reserve+push_back if you know the final size?

Bouncner replied:
Would you use `reserve()` + `[i]`? I thought we measured that both are equally fast? If not, I'll change it.

mrks replied:
That would be UB. `resize() + []` however, is significantly faster, because it does neither capacity checks nor has to update the actual size of the vector.

Bouncner replied:
Sorry, that's what I meant. We need an additional run variable though. Will change.

Bouncner replied:
Changed.

mrks commented on src/lib/storage/chunk_encoder.cpp:
```diff
@@ -12,41 +12,114 @@
+        }
+      });
+      std::cout << "WTF? " << values.size() << " & " << null_values.size()
+                << std::endl;  // testing what's wrong with the CI
```
Just making a comment so that this does not get lost

mrzzzrm commented:
Please don't forget the "Fix #..." in you opening post, so that the issue gets automatically closed when this PR is merged.

Bouncner commented:
I addressed most issues (the `check_table_equal` prints bother me and two smaller issues are open), but I struggle to the get the memcheck passed. I did a full leak check on Pella and anything went through just fine.
Has anybody an idea here?

Bouncner commented:
I addressed most issues (the `check_table_equal` prints bother me, the somewhat necessary `::opossum::`, and the shared pointer handling in the dictionary segment encoder). Any help and comments are welcome.

Bouncner commented:
WIP'ed the PR again. I am currently working on the calibration framework and recognized some unnecessary passing of encoding types and vector compression types (and partially the lack of vector compression types). As I saw this issue in other parts as well, I'll try to clean up the "encoding handling" a little bit.

Bouncner commented:
Ready to review again. Two small issues remain: I am not certain about the shared_pointer handling/allocation as discussed in an issue with @mrks and I am not certain about the output of `check_segments_equal()`.

mrks commented on src/lib/storage/base_segment_encoder.hpp:
```diff
@@ -131,14 +131,18 @@ class SegmentEncoder : public BaseSegmentEncoder {
 
-    const auto value_segment = std::dynamic_pointer_cast<const ValueSegment<ColumnDataType>>(base_value_segment);
-    Assert(value_segment, "Value segment must have passed data type.");
+    // Obtain allocator of segment to encode. Currently, only ValueSegment store the used allocator.
```
One `s` missing here - you decide which one ;)

mrks commented on src/lib/storage/chunk_encoder.cpp:
```diff
@@ -12,42 +12,118 @@
+      }
+    }
+
+    // In case of unencoded re-encoding, an any segment iterable is used to manually setup
```
Use the actual class name (`AnySegmentIterable`) so that it is easier to find should we ever want to rename it

mrks commented on src/lib/storage/chunk_encoder.hpp:
```diff
@@ -39,23 +28,21 @@ using ChunkEncodingSpec = std::vector<SegmentEncodingSpec>;
 
   /**
-   * @brief Encodes a chunk using the same segment-encoding spec
+   * @brief Encodes a chunk using the same segment encoding spec
```
Use the class name

mrks commented on src/lib/storage/dictionary_segment/dictionary_encoder.hpp:
```diff
@@ -28,112 +29,92 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
+                                                 const PolymorphicAllocator<T>& allocator) {
+    // Vectors to gather the input segment's data. This data is used in a later step to
+    // construct the actual dictionary and attribute vector.
+    std::vector<T> values;          // does contain the actual values (no NULLs)
```
Simply “contains”? Also, just to make sure no one misses it “This is different from other definitions of `values`”?

Bouncner replied:
Changed to "contains". I guess you mean that other times `values` stores NULL values as well? I think using `vector<T>` makes that clear enough. Or do you mean something else?

mrks replied:
(Almost?) everywhere else, this vector stores `T{}` for NULL values. Otherwise, we wouldn't be able to access values by index.

Here, it is fine, because we never access by index. I'd even go as far as to call it `dense_values`

mrks commented on src/lib/storage/dictionary_segment/dictionary_encoder.hpp:
```diff
@@ -28,112 +29,92 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
-      for (; value_it != values.cend(); ++value_it) {
-        const auto value_id = _get_value_id(dictionary, *value_it);
-        attribute_vector.push_back(value_id);
+  std::shared_ptr<BaseEncodedSegment> _encode_dictionary_segment(const U& dictionary, const std::vector<T>& values,
```
This only creates the attribute vector, but it does not create a dictionary segment from scratch. In itself, it seems useless. Move it into a lambda `create_attribute_vector` that is part of `_on_encode`?

Bouncner replied:
I adapted the function in a way that is is only responsible for the creation of compressed attribute vectors.
However, I am too stupid to move it into a simple lambda since I fail to get around the error, that dictionary template type cannot be deduced (_on_encode had two template parameters, T and S, and S was the parameter of the lambda function, S could not be deduced).

mrks commented on src/lib/storage/dictionary_segment/dictionary_encoder.hpp:
```diff
@@ -28,112 +29,92 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
 
     // We need to increment the dictionary size here because of possible null values.
-    const auto max_value = dictionary.size() + 1u;
+    const auto max_value = dictionary->size() + 1u;
```
max_value_id?

mrks commented on src/lib/storage/dictionary_segment/dictionary_encoder.hpp:
```diff
@@ -28,112 +29,92 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
 
     if constexpr (Encoding == EncodingType::FixedStringDictionary) {
-      return std::allocate_shared<FixedStringDictionarySegment<T>>(alloc, dictionary_sptr, attribute_vector_sptr,
+      return std::allocate_shared<FixedStringDictionarySegment<T>>(allocator, dictionary, encoded_attribute_vector,
```
Encoded vs compressed

mrks commented on src/lib/storage/lz4_segment/lz4_encoder.hpp:
```diff
@@ -63,32 +63,31 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
+      null_values.resize(segment_size);
+
+      // iterate over the segment to access the values and increment the row index to copy values and null flags
+      for (auto row_index = size_t{0u}; segment_it != segment_end; ++segment_it, ++row_index) {
```
Not a fan of non-classical for-loops. I find it too easy to miss something

mrks commented on src/lib/storage/lz4_segment/lz4_encoder.hpp:
```diff
@@ -63,32 +63,31 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
-        null_values[row_index] = segment_value.is_null();
-        contains_null_value = contains_null_value || segment_value.is_null();
+        null_values[row_index] = is_null;
+        contains_null_value = contains_null_value || is_null;
```
Simply `contains_null`?

mrks commented on src/lib/storage/lz4_segment/lz4_encoder.hpp:
```diff
@@ -168,37 +165,40 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
      */
-    auto sample_sizes = pmr_vector<size_t>{alloc};
-    sample_sizes.resize(num_elements);
+    auto string_samples_lengths = pmr_vector<size_t>{allocator};
```
I probably asked this when we originally brought this is, but how are these “samples” if we put in every single value?

Bouncner replied:
The values are(/might be) later "sampled" (taking the `n` first values) within `_train_dictionary(const pmr_vector<T>& values, const pmr_vector<size_t>& sample_sizes)`.

Bouncner replied:
There are several issues with the current LZ4 segment implementation, which I plan to rewrite. Shall I open an issue with an overview or do you trust me to address these issues in an upcoming PR?

Amongst these things are the removal of offsets, larger blocks (we basically never fill up the 4 KB), "real" sampling (tracking unique values and using the most often occuring values when samples too large), and more.

mrks commented on src/lib/storage/segment_encoding_utils.cpp:
```diff
@@ -40,16 +39,34 @@ std::unique_ptr<BaseSegmentEncoder> create_encoder(EncodingType encoding_type) {
 
-  if (zero_suppression_type) {
-    encoder->set_vector_compression(*zero_suppression_type);
+  if (encoder->uses_vector_compression() && encoding_spec.vector_compression_type) {
```
In which case would one of these be true and the other false?

Bouncner replied:
To be honest, I thought the check is not necessary, but valgrind complained and I was not able to reproduce it locally. The check fixed it for the CI setup. I will take another look at it.

mrks replied:
I'd be interested in the valgrind message

Bouncner replied:
Cannot reproduce. `encoder->uses_vector_compression()` has been removed again.

mrks commented on src/lib/utils/check_table_equal.cpp:
```diff
@@ -95,6 +95,33 @@ bool almost_equals(T left_val, T right_val, FloatComparisonMode float_comparison
+    return TableType::Data;
+  };
+
+  auto table_to_test = std::make_shared<Table>(definitions, table_type(segment_to_test));
```
I believe we call the inputs actual and <s>h...</s> expected

Bouncner replied:
Isn't actual the same as expected?

mrks replied:
No. See the ten of `actual_lqp / expected_lqp` in the tests. Also:

https://github.com/hyrise/hyrise/blob/0046cfdb7a6ad4937294cbca64f4e39f68558e0e/src/lib/utils/check_table_equal.cpp#L148-L151

mrks commented on src/lib/utils/check_table_equal.cpp:
```diff
@@ -95,6 +95,33 @@ bool almost_equals(T left_val, T right_val, FloatComparisonMode float_comparison
+  auto expected_table = std::make_shared<Table>(definitions, table_type(expected_segment));
+  expected_table->append_chunk(pmr_vector<std::shared_ptr<BaseSegment>>{expected_segment});
+
+  // The optional returned by check_table_equal is set when the function failed to determine equality.
```
Took me three times to understand. Maybe “If `check_table_equal` returns something other than std::nullopt, a difference has been found.”

mrks commented on src/lib/visualization/pqp_visualizer.cpp:
```diff
@@ -7,6 +7,7 @@
 #include "operators/limit.hpp"
 #include "operators/projection.hpp"
 #include "operators/table_scan.hpp"
+#include "utils/format_bytes.hpp"
```
?

Bouncner replied:
Removed it in another file which did not use any byte formatting. And the PQP visualizer uses format_bytes.

mrks commented on src/test/storage/any_segment_iterable_test.cpp:
```diff
@@ -83,6 +78,8 @@ INSTANTIATE_TEST_CASE_P(
-                      SegmentEncodingSpec{EncodingType::Dictionary, VectorCompressionType::SimdBp128}), );  // NOLINT
+                      SegmentEncodingSpec{EncodingType::Dictionary, VectorCompressionType::SimdBp128},
+                      SegmentEncodingSpec{EncodingType::FrameOfReference}, SegmentEncodingSpec{EncodingType::RunLength},
+                      SegmentEncodingSpec{EncodingType::LZ4}), );  // NOLINT
```
Argh. We have too many places where these definitions have to be added. Can we centralize this? Maybe as part of a different PR?

Bouncner replied:
I do not really know what you're meaning. For testing, I rather like to explicitly list all variants and especially test the non-default vector compressions.

Do you mean you'd like a function that yields all possible combinations?

mrks replied:
LZ4 should have been added here when it was originally introduced. The fact that it wasn't shows that we have too many places where new segment types have to be added. I'll open an issue.

Bouncner replied:
Ah yes, got you. Makes sense.

mrks commented on src/lib/utils/check_table_equal.hpp:
```diff
@@ -25,6 +25,10 @@ enum class TypeCmpMode { Strict, Lenient };
  */
 enum class FloatComparisonMode { RelativeDifference, AbsoluteDifference };
 
+bool check_segment_equal(const std::shared_ptr<BaseSegment>& segment_to_test,
```
Doc!

mrks commented on src/test/testing_assert.hpp:
```diff
@@ -64,6 +64,27 @@ bool contained_in_query_plan(const std::shared_ptr<const AbstractOperator>& node
+ * Compare two segments with respect to OrderSensitivity, TypeCmpMode and FloatComparisonMode
+ */
+#define EXPECT_SEGMENT_EQ(segment_to_test, expected_segment, order_sensitivity, type_cmp_mode, float_comparison_mode) \
+  EXPECT_TRUE(segment_to_test&& expected_segment&& check_segment_equal(                                               \
```
Looks like the formatter went berserk here

Bouncner replied:
It wouldn't bother me. Shall I add the two lines for clang-format to disable it or leave it as is?

mrks replied:
It took me some seconds to understand what happens here. I'd add those two lines.

Bouncner replied:
Wait? You are not talking about the missing 10 spaces for the third line but about something else, right? What did confuse you exactly?

mrks replied:
The `&&` that looks like a universal reference but is an "and"

mrks commented on src/lib/utils/check_table_equal.cpp:
```diff
@@ -113,12 +113,12 @@ bool check_segment_equal(const std::shared_ptr<BaseSegment>& segment_to_test,
 
-  auto table_to_test = std::make_shared<Table>(definitions, table_type(segment_to_test));
-  table_to_test->append_chunk(pmr_vector<std::shared_ptr<BaseSegment>>{segment_to_test});
+  auto table_to_test = std::make_shared<Table>(definitions, table_type(actual_segment));
```
actual_table

mrks commented on src/lib/storage/dictionary_segment/dictionary_encoder.hpp:
```diff
@@ -58,19 +58,48 @@ class DictionaryEncoder : public SegmentEncoder<DictionaryEncoder<Encoding>> {
 
+    const auto null_value_id = static_cast<uint32_t>(dictionary->size());
+
+    auto create_compressed_attribute_vector = [&](const auto& local_dictionary) {
```
Rename the dictionary from line 61 to `temp_dictionary` and use `dictionary` here?
